### PR TITLE
Fix minor datagram send bugs

### PR DESCRIPTION
**Pull request by Ralith**
Noticed while exploring solutions to #1832. Also tossing in any other simple wins I stumble across.

### CODE REVIEW

File: quinn-proto/src/connection/datagrams.rs [hunk 1]
```diff
@@ -68,11 +68,19 @@ impl<'a> Datagrams<'a> {
     ///
     /// Not necessarily the maximum size of received datagrams.
     pub fn max_size(&self) -> Option<usize> {
+        let key = self.conn.spaces[SpaceId::Data].crypto.as_ref().map_or_else(
```
Comment (djc) [inline quinn-proto/src/connection/datagrams.rs:71]: Would be nice to split the large expression in a separate commit and then fix the panic from there.

I suspect that, as in the other PR, `match` would be easier to follow than `map_or_else()`.

File: quinn-proto/src/connection/datagrams.rs [hunk 2]
```diff
@@ -149,10 +149,28 @@ impl DatagramState {
         Ok(was_empty)
     }
 
-    pub(super) fn write(&mut self, buf: &mut Vec<u8>, max_size: usize) -> bool {
-        let datagram = match self.outgoing.pop_front() {
-            Some(x) => x,
-            None => return false,
+    pub(super) fn write(
+        &mut self,
+        buf: &mut Vec<u8>,
```
Comment (djc) [inline quinn-proto/src/connection/datagrams.rs:154]: Can you add a docstring that elucidates the distinction between `max_size` and `drop_above_size`?

**Review by djc — APPROVED** [review]

File: quinn-proto/src/connection/datagrams.rs (hunk 2 above)
Comment (Ralith) [reply quinn-proto/src/connection/datagrams.rs:154]: I've instead rewritten this logic to proactively drop all oversized datagrams when a black hole is detected, rather than checking them lazily. It's a cosmetic difference for purposes of this PR, but it preempts a bug in #1837 where datagrams could become unrecoverably clogged by an over-MTU datagram due to `can_send_1rtt` skipping datagrams when an over-segment-size datagram is at the front of the queue, before packet assembly takes place.

File: quinn-proto/src/connection/datagrams.rs [hunk 3]
```diff
@@ -149,6 +149,27 @@ impl DatagramState {
         Ok(was_empty)
     }
 
+    /// Discard outgoing datagrams with a payload larger than `max_payload` bytes
+    ///
+    /// Used to ensure that reductions in MTU don't get us stuck in a state where we have a datagram
+    /// queued but can't send it.
+    pub(super) fn drop_oversized(&mut self, max_payload: usize) {
+        for i in (0..self.outgoing.len()).rev() {
```
Comment (djc) [inline quinn-proto/src/connection/datagrams.rs:157]: Couldn't this be a call to `self.outgoing.retain()`?

**Review by djc — APPROVED** [review]
Nice!

File: quinn-proto/src/connection/datagrams.rs (hunk 2 above)
Comment (djc) [reply quinn-proto/src/connection/datagrams.rs]: Makes sense.

File: quinn-proto/src/connection/datagrams.rs (hunk 3 above)
Comment (Ralith) [reply quinn-proto/src/connection/datagrams.rs]: Good catch.
